static/app/api/apiOptions.spec.ts

```diff
@@ -0,0 +1,164 @@
+        '/projects/$orgSlug/$projectSlug/releases/$releaseVersion/',
+        {
+          staleTime: 0,
+          path: {
+            orgSlug: 'my-org',
+            projectSlug: 'my-project',
+            releaseVersion: 'v1.0.0',
+          },
+        }
+      );
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<Release>>();
+    });
+
+    test('should not allow invalid path parameters', () => {
+      const options = apiOptions('/projects/$orgSlug/', {
+        staleTime: 0,
+        // @ts-expect-error Invalid path parameter
+        path: {orgSlug: 'my-org', invalidParam: 'invalid'},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should not allow excess path parameters', () => {
+      const options = apiOptions('/projects/$orgSlug/', {
+        staleTime: 0,
+        // @ts-expect-error Excess path parameter
+        path: {orgSlug: 'my-org', extraParam: 'extra'},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should require path params for paths with parameters', () => {
+      expect(() => {
+        const options = apiOptions('/projects/$orgSlug/', {
+          staleTime: 0,
+          // @ts-expect-error Missing required path parameter
+          path: {},
+        });
+
+        expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+      }).toThrow('Missing path param: orgSlug');
+    });
+
+    test('should not allow empty path parameters for paths without parameters', () => {
+      const options = apiOptions('/projects/', {
+        staleTime: 0,
+        // @ts-expect-error Empty path parameters not allowed
+        path: {},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should not need path params for paths without parameters', () => {
+      const options = apiOptions('/projects/', {
+        staleTime: 0,
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should allow string or number path parameters', () => {
+      const options = apiOptions('/items/$id/', {
+        staleTime: 0,
+        path: {id: 123}, // number
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+
+      const options2 = apiOptions('/items/$id/', {
+        staleTime: 0,
+        path: {id: 'abc'}, // string
+      });
+
+      expectTypeOf(options2.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should default to never for unknown API paths', () => {
+      const options = apiOptions('/unknown/$param/', {
+        staleTime: 0,
+        path: {param: 'value'},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
```

Comment:
if you pass in a path that is not defined in `ApiDefinition`, the `queryFn` will return `never`. We could also make it `unknown`, but this doesn’t really matter as it’s not encouraged. If you want to explicitly type the return value, you need to use the `.ReturnType` builder pattern.